yb-voyager/cmd/eventQueue.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright (c) YugabyteDB, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/yugabyte/yb-voyager/yb-voyager/src/tgtdb"
+	"github.com/yugabyte/yb-voyager/yb-voyager/src/utils"
+)
+
+const (
+	QUEUE_DIR_NAME               = "queue"
+	QUEUE_SEGMENT_FILE_NAME      = "segment"
+	QUEUE_SEGMENT_FILE_EXTENSION = "ndjson"
+)
+
+type EventQueue struct {
+	QueueDirPath       string
+	SegmentNumToStream int64
+}
+
+func NewEventQueue(exportDir string) *EventQueue {
+	return &EventQueue{
+		QueueDirPath:       filepath.Join(exportDir, "data", QUEUE_DIR_NAME),
+		SegmentNumToStream: 0,
+	}
+}
+
+// GetNextSegment returns the next segment to process
+func (eq *EventQueue) GetNextSegment() (*EventQueueSegment, error) {
+	log.Infof("getting next segment to process from %s", eq.QueueDirPath)
+
+	segmentFilePath := filepath.Join(eq.QueueDirPath, fmt.Sprintf("%s.%d.%s", QUEUE_SEGMENT_FILE_NAME, eq.SegmentNumToStream, QUEUE_SEGMENT_FILE_EXTENSION))
+	_, err := os.Stat(segmentFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get next segment file path: %w", err)
+	}
+
+	segment := NewEventQueueSegment(segmentFilePath, eq.SegmentNumToStream)
+	eq.SegmentNumToStream++
+	return segment, nil
+}
+
+type EventQueueSegment struct {
+	FilePath   string
+	SegmentNum int64 // 0-based
+	processed  bool
+	file       *os.File
+	scanner    *bufio.Scanner
+	buffer     []byte // buffer for scanning from file
+}
+
+var EOFMarker = `\.`
+
+func NewEventQueueSegment(filePath string, segmentNum int64) *EventQueueSegment {
+	return &EventQueueSegment{
+		FilePath:   filePath,
+		SegmentNum: segmentNum,
+		processed:  false,
+	}
+}
+
+func (eqs *EventQueueSegment) Open() error {
+	file, err := os.OpenFile(eqs.FilePath, os.O_RDONLY, 0640)
+	if err != nil {
+		return fmt.Errorf("failed to open segment file %s: %w", eqs.FilePath, err)
+	}
+	eqs.file = file
+	eqs.scanner = bufio.NewScanner(utils.NewTailReader(file))
+
+	// providing buffer to scanner for scanning
+	eqs.buffer = make([]byte, 0, 100*ONE_KB)
```

Comment:
I am worried about the kind of huge multi MB records that we saw in blockjoy testing.
What happens if line doesn't fit in 100KB.

Reply:
yes, created a ticket for that: https://yugabyte.atlassian.net/browse/DB-7318